File: product/dashboard/src/test/java/io/codekvast/dashboard/file_import/FileImportTaskTest.java
package io.codekvast.dashboard.file_import;

import io.codekvast.dashboard.bootstrap.CodekvastDashboardSettings;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.io.File;
import java.io.IOException;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

/**
 * @author olle.hallin@crisp.se
 */
@RunWith(MockitoJUnitRunner.class)
public class FileImportTaskTest {

    @Rule
    public TemporaryFolder temporaryFolder = new TemporaryFolder();

    @Mock
    private PublicationImporter importer;

    private CodekvastDashboardSettings settings;

    private FileImportTask task;

    @Before
    public void beforeTest() {
        settings = new CodekvastDashboardSettings();
        settings.setQueuePath(temporaryFolder.getRoot());

        task = new FileImportTask(settings, importer);
    }

    @Test
    public void should_handle_empty_queuePath() {
        // given
        // An empty queuePath

        // when
        task.importPublicationFiles();

        // then
        verifyNoMoreInteractions(importer);
    }

    @Test
    public void should_ignore_non_zip_files() throws Exception {
        // given
        File file = createImportFile(".bar");

        // when
        task.importPublicationFiles();

        // then
        assertThat(file.exists(), is(true));
        verifyNoMoreInteractions(importer);
    }

    @Test
    public void should_delete_file_after_successful_import() throws Exception {
        // given
        settings.setDeleteImportedFiles(true);
        File file = createImportFile(".ser");
        when(importer.importPublicationFile(any(File.class))).thenReturn(true);

        // when
        task.importPublicationFiles();

        // then
        verify(importer).importPublicationFile(file);
        assertThat(file.exists(), is(false));
    }

    @Test
    public void should_not_delete_file_after_failed_import() throws Exception {
        // given
        settings.setDeleteImportedFiles(true);
        File file = createImportFile(".ser");
        when(importer.importPublicationFile(any(File.class))).thenReturn(false);

        // when
        task.importPublicationFiles();

        // then
        verify(importer).importPublicationFile(file);
        assertThat(file.exists(), is(true));
    }

    @Test
    public void should_not_delete_file_after_import() throws Exception {
        // given
        settings.setDeleteImportedFiles(false);
        File file = createImportFile(".ser");

        // when
        task.importPublicationFiles();

        // then
        verify(importer).importPublicationFile(file);
        assertThat(file.exists(), is(true));
    }

    private File createImportFile(String suffix) throws IOException {
        return File.createTempFile("import", suffix, temporaryFolder.newFolder());
    }

}

File: product/server/intake/src/test/kotlin/io/codekvast/intake/file_import/FileImportTaskTest.kt
package io.codekvast.intake.file_import

import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import io.codekvast.common.lock.Lock
import io.codekvast.common.lock.LockManager
import io.codekvast.common.lock.LockTemplate
import io.codekvast.intake.bootstrap.CodekvastIntakeSettings
import org.hamcrest.MatcherAssert.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoMoreInteractions
import org.mockito.MockitoAnnotations
import java.io.File
import java.util.*
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import org.hamcrest.Matchers.`is` as is_

/** @author olle.hallin@crisp.se
 */
class FileImportTaskTest {
    @TempDir
    lateinit var temporaryFolder: File

    @Mock
    private lateinit var importer: PublicationImporter

    @Mock
    private lateinit var lockManager: LockManager

    @BeforeEach
    fun beforeTest() {
        MockitoAnnotations.openMocks(this)
        whenever(lockManager.acquireLock(ArgumentMatchers.any()))
            .thenReturn(Optional.of(Lock.forTask("test", 60)))
    }

    private fun createFileImportTask(deleteImportedFiles: Boolean): FileImportTask {
        val settings = CodekvastIntakeSettings(
            fileImportQueuePath = temporaryFolder,
            deleteImportedFiles = deleteImportedFiles
        )
        return FileImportTask(settings, importer, mock(), LockTemplate(lockManager))
    }

    @Test
    fun should_handle_empty_fileImportQueuePath() {
        // given
        // An empty fileImportQueuePath
        val task = createFileImportTask(true)

        // when
        task.importPublicationFiles()

        // then
        verifyNoMoreInteractions(importer)
    }

    @Test
    fun should_ignore_non_ser_files() {
        // given
        val file = createImportFile(".bar")
        val task = createFileImportTask(true)

        // when
        task.importPublicationFiles()

        // then
        assertTrue(file.exists())
        verifyNoMoreInteractions(importer)
    }

    @Test
    fun should_delete_file_after_successful_import() {
        // given
        whenever(importer.importPublicationFile(any())).thenReturn(true)
        val file = createImportFile(".ser")
        val task = createFileImportTask(true)

        // when
        task.importPublicationFiles()

        // then
        verify(importer).importPublicationFile(file)
        assertFalse(file.exists())
    }

    @Test
    fun should_not_delete_file_after_failed_import() {
        // given
        whenever(importer.importPublicationFile(any())).thenReturn(false)
        val file = createImportFile(".ser")
        val task = createFileImportTask(true)

        // when
        task.importPublicationFiles()

        // then
        verify(importer).importPublicationFile(file)
        assertTrue(file.exists())
    }

    @Test
    fun should_not_delete_file_after_import() {
        // given
        val file = createImportFile(".ser")
        val task = createFileImportTask(false)

        // when
        task.importPublicationFiles()

        // then
        verify(importer).importPublicationFile(file)
        assertTrue(file.exists())
    }

    private fun createImportFile(suffix: String): File {
        return File.createTempFile("import-", suffix, temporaryFolder)
    }
}